engine/runtime-integration-tests/src/test/java/org/enso/compiler/TypeInferenceTest.java

```diff
@@ -0,0 +1,1192 @@
+                """
+                    f x =
+                      y = case x of
+                        1 -> x
+                        "foo" -> x
+                      y
+                    """,
+                uri.getAuthority())
+            .uri(uri)
+            .buildLiteral();
+
+    var module = compile(src);
+    var f = findStaticMethod(module, "f");
+    assertSumType(findAssignment(f, "y").expression(), "Integer", "Text");
+  }
+
+  @Ignore("TODO")
+  @Test
+  public void inferEqualityBoundsFromCaseEdgeCase() throws Exception {
+    // This test ensures that the equality bound from _:Other_Type is only applicable in its branch
+    // and does not 'leak' to other branches.
+    final URI uri = new URI("memory://inferEqualityBoundsFromCaseEdgeCase.enso");
+    final Source src =
+        Source.newBuilder(
+                "enso",
+                """
+                    type My_Type
+                        Value v
+                    type Other_Type
+                        Value o
+                    f x =
+                      y = case x of
+                        _ : Other_Type -> My_Type.Value 42
+                        _ : My_Type -> x
+                      y
+                    """,
+                uri.getAuthority())
+            .uri(uri)
+            .buildLiteral();
+
+    var module = compile(src);
+    var myType = "inferEqualityBoundsFromCaseEdgeCase.My_Type";
+    var f = findStaticMethod(module, "f");
+    assertAtomType(myType, findAssignment(f, "y").expression());
+  }
+
+  @Test
+  public void sumTypeFromCase() throws Exception {
+    final URI uri = new URI("memory://sumTypeFromCase.enso");
+    final Source src =
+        Source.newBuilder(
+                "enso",
+                """
+                    type My_Type
+                        Value v
+                    type Other_Type
+                        Value o
+                    f x =
+                      y = case x of
+                        1 -> My_Type.Value 42
+                        2 -> Other_Type.Value 23
+                      y
+                    """,
+                uri.getAuthority())
+            .uri(uri)
+            .buildLiteral();
+
+    var module = compile(src);
+    var f = findStaticMethod(module, "f");
+    assertSumType(findAssignment(f, "y").expression(), "My_Type", "Other_Type");
+  }
+
+  @Ignore
+  @Test
+  public void sumTypeFromIf() throws Exception {
+    final URI uri = new URI("memory://sumTypeFromIf.enso");
+    final Source src =
+        Source.newBuilder(
+                "enso",
+                """
+                    f x =
+                      y = if x == 1 then "foo" else 42
+                      y
+                    """,
+                uri.getAuthority())
+            .uri(uri)
+            .buildLiteral();
+
+    var module = compile(src);
+    var f = findStaticMethod(module, "f");
+    assertSumType(findAssignment(f, "y").expression(), "Text", "Integer");
+  }
+
+  @Ignore
```

Comment:
Exactly. I thought they will get special handling in this iteration, like `case` got; but since they are actually methods, it was a bit 'too special' for that - I will try handling `if` together with regular methods in the second iteration that I will be starting soon.